File: features/authApi.ts
import type {AuthSession, AuthUser, LanguagePreference} from '../entities';
import {CURRENT_CONSENT_VERSION} from '../entities';
import {apiRequest} from './apiClient';

export type RegisterInput = {
    email: string;
    password: string;
    language?: LanguagePreference;
    acceptedTerms: boolean;
    acceptedPrivacy: boolean;
    acceptedPersonalData: boolean;
};

export type VerificationStart = {
    email: string;
    verificationRequired: boolean;
};

export const loginRequest = (email: string, password: string) =>
    apiRequest<AuthSession>('/api/v1/auth/login', {
        method: 'POST',
        auth: false,
        body: {email, password},
    });

export const registerRequest = (input: RegisterInput) =>
    apiRequest<VerificationStart>('/api/v1/auth/register', {
        method: 'POST',
        auth: false,
        body: {
            email: input.email,
            password: input.password,
            language: input.language,
            acceptedTerms: input.acceptedTerms,
            acceptedPrivacy: input.acceptedPrivacy,
            acceptedPersonalData: input.acceptedPersonalData,
            consentVersion: CURRENT_CONSENT_VERSION,
        },
    });

export const verifyEmailRequest = (email: string, code: string) =>
    apiRequest<AuthSession>('/api/v1/auth/verify-email', {
        method: 'POST',
        auth: false,
        body: {email, code},
    });

export const resendVerificationRequest = (email: string, language?: LanguagePreference) =>
    apiRequest<{ok: boolean}>('/api/v1/auth/resend-verification', {
        method: 'POST',
        auth: false,
        body: {email, language},
    });

export const forgotPasswordRequest = (email: string, language?: LanguagePreference) =>
    apiRequest<{ok: boolean}>('/api/v1/auth/forgot-password', {
        method: 'POST',
        auth: false,
        body: {email, language},
    });

export const resetPasswordRequest = (email: string, code: string, password: string) =>
    apiRequest<AuthSession>('/api/v1/auth/reset-password', {
        method: 'POST',
        auth: false,
        body: {email, code, password},
    });

export const fetchMe = () => apiRequest<AuthUser>('/api/v1/me');

export const logoutRequest = (refreshToken: string) =>
    apiRequest<{ok: boolean}>('/api/v1/auth/logout', {
        method: 'POST',
        auth: false,
        body: {refreshToken},
    });

export const deleteAccountRequest = () =>
    apiRequest<{ok: boolean}>('/api/v1/me', {
        method: 'DELETE',
    });

File: ios-swift/MoonCompass/Core/Networking/Endpoints/AuthAPI.swift
import Foundation

/// Account endpoints. Every response that carries a token pair is persisted
/// here, so screens only ever see the `User`.
struct AuthAPI: Sendable {
    let client: APIClient

    /// Creates the account and sends a verification code. No tokens yet — the
    /// session starts at `verifyEmail`.
    func register(_ request: RegisterRequest) async throws -> RegisterResponse {
        try await client.send(.post, "/api/v1/auth/register", body: request, auth: .none)
    }

    /// Throws `invalid_credentials` (401) or `email_unverified` (403).
    func login(email: String, password: String) async throws -> User {
        try await establishSession(
            client.send(.post, "/api/v1/auth/login", body: LoginRequest(email: email, password: password), auth: .none)
        )
    }

    func verifyEmail(email: String, code: String) async throws -> User {
        try await establishSession(
            client.send(.post, "/api/v1/auth/verify-email", body: VerifyEmailRequest(email: email, code: code), auth: .none)
        )
    }

    /// Throws `code_cooldown` (429, with `retryAfter`) when asked too soon.
    func resendVerification(email: String, language: String) async throws {
        try await client.sendIgnoringResponse(
            .post, "/api/v1/auth/resend-verification",
            body: EmailLanguageRequest(email: email, language: language), auth: .none
        )
    }

    func forgotPassword(email: String, language: String) async throws {
        try await client.sendIgnoringResponse(
            .post, "/api/v1/auth/forgot-password",
            body: EmailLanguageRequest(email: email, language: language), auth: .none
        )
    }

    func resetPassword(email: String, code: String, password: String) async throws -> User {
        try await establishSession(
            client.send(
                .post, "/api/v1/auth/reset-password",
                body: ResetPasswordRequest(email: email, code: code, password: password), auth: .none
            )
        )
    }

    /// Always ends the local session, even if the server cannot be reached: the
    /// user asked to sign out, and an unrevoked refresh token simply expires.
    func logout() async {
        if let refreshToken = await client.currentRefreshToken() {
            try? await client.sendIgnoringResponse(
                .post, "/api/v1/auth/logout",
                body: RefreshTokenRequest(refreshToken: refreshToken), auth: .none
            )
        }
        await client.clearSession()
    }

    func me() async throws -> User {
        try await client.send(.get, "/api/v1/me", auth: .required)
    }

    /// Clears the local session only once the server confirmed the deletion, so
    /// a failed request leaves the user signed in and able to retry.
    func deleteAccount() async throws {
        try await client.sendIgnoringResponse(.delete, "/api/v1/me", auth: .required)
        await client.clearSession()
    }

    private func establishSession(_ response: AuthResponse) async throws -> User {
        try await client.storeSession(response.tokens)
        return response.user
    }
}
